python/tchannel/dispatch.py

```diff
@@ -56,21 +56,32 @@ def _call_endpoint(self, endpoint, request, response):
         finally:
             response.finish()
 
-    def handle_call(self, message_id, call, connection):
-        request = Request(message_id, call, connection)
-        endpoint = self.endpoints.get(request.endpoint, None)
+    @tornado.gen.coroutine
+    def handle_call(self, request, connection):
+        # read arg_1 so that handle_call is able to get the endpoint
+        # name and find the endpoint handler.
+        # the arg_1 value will be store in the request.endpoint field.
 
+        # NOTE: after here, the correct way to access value of arg_1 is through
+        # request.endpoint. The original argstream[0] is no longer valid. If
+        # user still tries read from it, it will return empty.
+        chunk = yield request.argstreams[0].read()
+        while chunk:
+            request.endpoint += chunk
+            chunk = yield request.argstreams[0].read()
+
+        endpoint = self.endpoints.get(request.endpoint, None)
         if endpoint is None:
-            return connection.senderror(
+            connection.senderror(
```

Comment:
I would leave the `return`s here. That way if `senderror` or `call_endpoint` return a future (which they most likely do), that's propagated up to the caller.